src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+        @Override
+        public void onAdd(Endpoints endpoints) {
+            logger.infov(
+                    "Endpoint {0} created in namespace {1}",
+                    endpoints.getMetadata().getName(), endpoints.getMetadata().getNamespace());
+            handleEndpointEvent(endpoints.getMetadata().getNamespace());
+        }
+
+        @Override
+        public void onUpdate(Endpoints oldEndpoints, Endpoints newEndpoints) {
+            logger.infov(
+                    "Endpoint {0} modified in namespace {1}",
+                    newEndpoints.getMetadata().getName(),
+                    newEndpoints.getMetadata().getNamespace());
+            handleEndpointEvent(newEndpoints.getMetadata().getNamespace());
+        }
+
+        @Override
+        public void onDelete(Endpoints endpoints, boolean deletedFinalStateUnknown) {
+            logger.infov(
+                    "Endpoint {0} deleted in namespace {1}",
+                    endpoints.getMetadata().getName(), endpoints.getMetadata().getNamespace());
+            if (deletedFinalStateUnknown) {
+                logger.warnv("Deleted final state unknown: {0}", endpoints);
+                return;
+            }
+            handleEndpointEvent(endpoints.getMetadata().getNamespace());
+        }
+    }
+
+    List<TargetTuple> tuplesFromEndpoints(Endpoints endpoints) {
+        List<TargetTuple> tts = new ArrayList<>();
+        for (EndpointSubset subset : endpoints.getSubsets()) {
+            for (EndpointPort port : subset.getPorts()) {
+                for (EndpointAddress addr : subset.getAddresses()) {
+                    tts.add(new TargetTuple(addr.getTargetRef(), addr, port));
+                }
+            }
+        }
+        return tts;
+    }
+
+    private class TargetTuple {
+        ObjectReference objRef;
+        EndpointAddress addr;
+        EndpointPort port;
+
+        TargetTuple(ObjectReference objRef, EndpointAddress addr, EndpointPort port) {
+            this.objRef = objRef;
+            this.addr = addr;
+            this.port = port;
+        }
+
+        public Target toTarget() {
+            try {
+                String ip = addr.getIp().replaceAll("\\.", "-");
+                String namespace = objRef.getNamespace();
+
+                boolean isPod = objRef.getKind().equals(KubeDiscoveryNodeType.POD.getKind());
+
+                String host = String.format("%s.%s", ip, namespace);
+                if (isPod) {
+                    host = String.format("%s.pod", host);
+                }
+
+                JMXServiceURL jmxUrl =
+                        new JMXServiceURL(
+                                "rmi",
+                                "",
+                                0,
+                                "/jndi/rmi://" + host + ':' + port.getPort() + "/jmxrmi");
+                URI connectUrl = URI.create(jmxUrl.toString());
+
+                try {
+                    return Target.getTargetByConnectUrl(connectUrl);
```

Comment:
I think it would be extremely uncommon, but this could yield an unexpected result if there is a URL collision between KubeApiDiscovery and another discovery mechanism, and the other mechanism has created a Target with this URL first. It seems like it would become possible for the Endpoints event handler to end up with this query getting executed and pulling a Target from a different Realm out of the database and returning it.

So I think the Target either needs to be found by traversing this Realm's subtree, or else it can be queried directly from the database like this but then the Realm subtree membership needs to be verified before returning it.

Reply:
Right! I think that makes sense! I can verify its associated node if its of type `Endpoint`.

Reply:
That works too. If it happens then I would say log a warning and return null, I guess? Or maybe just throw an `IllegalStateException` since this should be a very rare occurrence and only if there is a URL collision like that, which really means that there are overlapping Realms. Until https://github.com/cryostatio/cryostat3/issues/71 is solved, overlapping Realms should be considered an invalid configuration.